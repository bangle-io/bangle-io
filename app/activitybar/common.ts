import type React from 'react';

import { vars } from '@bangle.io/css-vars';

export const ButtonStyleOBj: React.CSSProperties = {
  borderRadius: 0,
  color: vars.color.app.activitybarText,
  padding: 0,
};
