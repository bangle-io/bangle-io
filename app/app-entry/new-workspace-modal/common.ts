export const FILE_SYSTEM = 'file-system';
export const BROWSER = 'browser';
export type WorkspaceStorageType = 'file-system' | 'browser';

export const ERROR_PICKING_DIRECTORY_ERROR = 'ERROR_PICKING_DIRECTORY';
export const UNKNOWN_ERROR = 'UNKNOWN_ERROR';
export const INVALID_WORKSPACE_NAME_ERROR = 'INVALID_WORKSPACE_NAME';
export const WORKSPACE_AUTH_REJECTED_ERROR = 'WORKSPACE_AUTH_REJECTED';
export const WORKSPACE_NAME_ALREADY_EXISTS_ERROR =
  'WORKSPACE_NAME_ALREADY_EXISTS';

export type WorkspaceCreateErrorTypes =
  | typeof ERROR_PICKING_DIRECTORY_ERROR
  | typeof INVALID_WORKSPACE_NAME_ERROR
  | typeof WORKSPACE_AUTH_REJECTED_ERROR
  | typeof WORKSPACE_NAME_ALREADY_EXISTS_ERROR
  | typeof UNKNOWN_ERROR;

export const BROWSE_BUTTON_ID = 'new-workspace-modal_browse-button';
export const CREATE_BUTTON_ID = 'new-workspace-modal_create-button';

export function getStorageDescription(storageType: WorkspaceStorageType) {
  switch (storageType) {
    case 'file-system': {
      return 'File system';
    }
    case 'browser': {
      return 'Browser';
    }
    default: {
      // hack to catch switch slipping
      let val: never = storageType;
      throw new Error('Unknown storage type ' + val);
    }
  }
}
