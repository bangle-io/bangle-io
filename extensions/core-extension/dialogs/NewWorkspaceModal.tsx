import React, { useEffect } from 'react';

import { nsmApi2 } from '@bangle.io/api';
import {
  NEW_BROWSER_WORKSPACE_DIALOG_NAME,
  NEW_NATIVE_FS_WORKSPACE_DIALOG_NAME,
  NEW_PRIVATE_FS_WORKSPACE_DIALOG_NAME,
  WorkspaceType,
} from '@bangle.io/constants';
import { useExtensionRegistryContext } from '@bangle.io/extension-registry';
import type { DialogComponentType } from '@bangle.io/shared-types';
import { TriState } from '@bangle.io/tri-state';
import { safeNavigatorStorageGetDirectory } from '@bangle.io/utils';

import { PickWorkspaceType } from './PickWorkspaceType';

export const NewWorkspaceModal: DialogComponentType = (props) => {
  const [isPrivateFsAvailable, setIsPrivateFsAvailable] =
    React.useState<TriState>(TriState.Unknown);

  useEffect(() => {
    let unmounted = false;

    safeNavigatorStorageGetDirectory().then((dir) => {
      if (unmounted) {
        return;
      }

      // setIsPrivateFsAvailable(TriState.Yes);
      // TODO private fs become buggy in safari for some reason
      setIsPrivateFsAvailable(TriState.No);
    });

    return () => {
      unmounted = true;
    };
  }, []);

  const extensionRegistry = useExtensionRegistryContext();

  if (isPrivateFsAvailable === TriState.Unknown) {
    return null;
  }

  return (
    <PickWorkspaceType
      hasGithub={extensionRegistry.isExtensionDefined(
        '@bangle.io/github-storage',
      )}
      hasPrivateFs={isPrivateFsAvailable === TriState.Yes}
      onDismiss={() => {
        props.onDismiss(props.dialogName);
      }}
      onSelect={(storage) => {
        switch (storage) {
          case WorkspaceType.Github: {
            nsmApi2.ui.showDialog({
              dialogName:
                'dialog::@bangle.io/github-storage:NEW_GITHUB_WORKSPACE_TOKEN_DIALOG',
            });

            return;
          }
          case WorkspaceType.Browser: {
            nsmApi2.ui.showDialog({
              dialogName: NEW_BROWSER_WORKSPACE_DIALOG_NAME,
            });

            return;
          }
          case WorkspaceType.NativeFS: {
            nsmApi2.ui.showDialog({
              dialogName: NEW_NATIVE_FS_WORKSPACE_DIALOG_NAME,
            });

            return;
          }

          case WorkspaceType.PrivateFS: {
            nsmApi2.ui.showDialog({
              dialogName: NEW_PRIVATE_FS_WORKSPACE_DIALOG_NAME,
            });

            return;
          }

          case WorkspaceType.Help: {
            throw new Error('Not allowed to create help workspace');

            return;
          }

          default: {
            let val: never = storage;
            throw new Error('Unknown storage type');
          }
        }
      }}
    />
  );
};
