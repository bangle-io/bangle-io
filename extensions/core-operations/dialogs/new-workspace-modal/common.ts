import { supportsNativeBrowserFs } from '@bangle.io/baby-fs';

export const FILE_SYSTEM = 'file-system';
export const BROWSER = 'browser';
export type WorkspaceStorageType = typeof FILE_SYSTEM | typeof BROWSER;

export const ERROR_PICKING_DIRECTORY_ERROR = 'ERROR_PICKING_DIRECTORY';
export const UNKNOWN_ERROR = 'UNKNOWN_ERROR';
export const INVALID_WORKSPACE_NAME_ERROR = 'INVALID_WORKSPACE_NAME';
export const WORKSPACE_AUTH_REJECTED_ERROR = 'WORKSPACE_AUTH_REJECTED';
// this happens when the chrome just rejects by saying user activation needed
export const CLICKED_TOO_SOON_ERROR = 'CLICKED_TOO_SOON_ERROR';
export const WORKSPACE_NAME_ALREADY_EXISTS_ERROR =
  'WORKSPACE_NAME_ALREADY_EXISTS';

export type WorkspaceCreateErrorTypes =
  | typeof ERROR_PICKING_DIRECTORY_ERROR
  | typeof INVALID_WORKSPACE_NAME_ERROR
  | typeof WORKSPACE_AUTH_REJECTED_ERROR
  | typeof WORKSPACE_NAME_ALREADY_EXISTS_ERROR
  | typeof CLICKED_TOO_SOON_ERROR
  | typeof UNKNOWN_ERROR;

export const BROWSE_BUTTON_ID = 'new-workspace-modal_browse-button';
export const CREATE_BUTTON_ID = 'new-workspace-modal_create-button';

export function getStorageDescription(storageType: WorkspaceStorageType) {
  switch (storageType) {
    case FILE_SYSTEM: {
      return 'File system';
    }
    case BROWSER: {
      return 'Browser';
    }
    default: {
      // hack to catch switch slipping
      let val: never = storageType;
      throw new Error('Unknown storage type ' + val);
    }
  }
}

export const defaultStorageType = supportsNativeBrowserFs()
  ? FILE_SYSTEM
  : BROWSER;
