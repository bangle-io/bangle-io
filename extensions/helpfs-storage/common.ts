export const extensionName = '@bangle.io/helpfs-storage';
