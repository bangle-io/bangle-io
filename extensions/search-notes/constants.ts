import type { SearchResultItem } from '@bangle.io/search-pm-node';

export const CONCURRENCY = 10;

export const SHOW_SEARCH_SIDEBAR_OPERATION =
  'operation::@bangle.io/search-notes:show-search-sidebar';

export const SEARCH_SIDEBAR_NAME =
  'sidebar::@bangle.io/search-notes:search-notes';
export const extensionName = '@bangle.io/search-notes';
export interface SearchNotesExtensionState {
  searchQuery: string;
  pendingSearch: boolean;
  searchResults: SearchResultItem[] | null;
  // a counter keeping track of external change to the search query
  // useful for resetting any local state
  externalInputChange: number;
}

export type SearchNotesActions =
  | {
      name: `action::@bangle.io/search-notes:update-state`;
      value: Partial<SearchNotesExtensionState>;
    }
  | {
      name: 'action::@bangle.io/search-notes:input-search-query';
      value: {
        query: string;
      };
    }
  | {
      name: 'action::@bangle.io/search-notes:external-search-query-update';
      value: {
        query: string;
      };
    };
