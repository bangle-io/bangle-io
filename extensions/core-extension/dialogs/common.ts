import { supportsNativeBrowserFs } from '@bangle.io/baby-fs';
import { WorkspaceType } from '@bangle.io/constants';

let nativeFsSupport = supportsNativeBrowserFs();

export const disabledStorageType: WorkspaceType[] = [
  !nativeFsSupport && WorkspaceType.NativeFS,
].filter((r): r is WorkspaceType => Boolean(r));

export const defaultStorage = nativeFsSupport
  ? WorkspaceType.NativeFS
  : WorkspaceType.Browser;

export const ERROR_PICKING_DIRECTORY_ERROR = 'ERROR_PICKING_DIRECTORY';
export const UNKNOWN_ERROR = 'UNKNOWN_ERROR';
export const INVALID_WORKSPACE_NAME_ERROR = 'INVALID_WORKSPACE_NAME';
export const WORKSPACE_AUTH_REJECTED_ERROR = 'WORKSPACE_AUTH_REJECTED';
// this happens when the chrome just rejects by saying user activation needed
export const CLICKED_TOO_SOON_ERROR = 'CLICKED_TOO_SOON_ERROR';
export const WORKSPACE_NAME_ALREADY_EXISTS_ERROR =
  'WORKSPACE_NAME_ALREADY_EXISTS';

export type WorkspaceCreateErrorTypes =
  | typeof ERROR_PICKING_DIRECTORY_ERROR
  | typeof INVALID_WORKSPACE_NAME_ERROR
  | typeof WORKSPACE_AUTH_REJECTED_ERROR
  | typeof WORKSPACE_NAME_ALREADY_EXISTS_ERROR
  | typeof CLICKED_TOO_SOON_ERROR
  | typeof UNKNOWN_ERROR;

export const BROWSE_BUTTON_ID = 'new-workspace-modal_browse-button';
