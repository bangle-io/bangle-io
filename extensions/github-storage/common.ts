import { nsmApi2 } from '@bangle.io/api';
import type { Severity } from '@bangle.io/constants';
import { SEVERITY, WorkspaceType } from '@bangle.io/constants';
import { acquireLockIfAvailable, isMobile } from '@bangle.io/utils';

export const EXTENSION_NAME = '@bangle.io/github-storage';

export const GITHUB_STORAGE_PROVIDER_NAME = WorkspaceType.Github;

export const OPERATION_UPDATE_GITHUB_TOKEN =
  'operation::@bangle.io/github-storage:update-github-token';

export const OPERATION_SYNC_GITHUB_CHANGES =
  'operation::@bangle.io/github-storage:sync-github-changes';

export const OPERATION_DISCARD_LOCAL_CHANGES =
  'operation::@bangle.io/github-storage:OPERATION_DISCARD_LOCAL_CHANGES';

export const OPERATION_SHOW_CONFLICT_DIALOG =
  'operation::@bangle.io/github-storage:show-conflict-dialog';

export const OPERATION_OPTIMIZE_GITHUB_STORAGE =
  'operation::@bangle.io/github-storage:OPERATION_OPTIMIZE_GITHUB_STORAGE';

export const DISCARD_LOCAL_CHANGES_DIALOG =
  'dialog::@bangle.io/github-storage:discard-local-changes';

export const NEW_GITHUB_WORKSPACE_TOKEN_DIALOG =
  'dialog::@bangle.io/github-storage:NEW_GITHUB_WORKSPACE_TOKEN_DIALOG';

export const NEW_GITHUB_WORKSPACE_REPO_PICKER_DIALOG =
  'dialog::@bangle.io/github-storage:NEW_GITHUB_WORKSPACE_REPO_PICKER_DIALOG';

export const UPDATE_GITHUB_TOKEN_DIALOG =
  'dialog::@bangle.io/github-storage:UPDATE_GITHUB_TOKEN_DIALOG';

export const CONFLICT_DIALOG = 'dialog::@bangle.io/github-storage:conflict';

export const LOCK_NAME = '@bangle.io/github-storage:sync-lock';

export interface GithubWsMetadata {
  owner: string;
  branch: string;
}

export async function getGithubSyncLockWrapper<
  R extends (...args: any[]) => Promise<any>,
>(
  wsName: string,
  cb: R,
): Promise<
  | {
      lockAcquired: true;
      result: Awaited<ReturnType<R>>;
    }
  | {
      lockAcquired: false;
      result: undefined;
    }
> {
  let releaseLock = await acquireLockIfAvailable(LOCK_NAME + ':' + wsName);

  if (!releaseLock) {
    return { lockAcquired: false, result: undefined };
  }

  try {
    let result = await cb();

    return {
      lockAcquired: true,
      result,
    };
  } finally {
    await releaseLock();
  }
}

export const getSyncInterval = () => (isMobile ? 15 * 1000 : 1000 * 60);

export function notify(title: string, severity: Severity, content?: string) {
  return nsmApi2.ui.showNotification({
    severity,
    title,
    uid: 'sync notification-' + Math.random(),
    transient: severity !== SEVERITY.ERROR,
    content,
  });
}
