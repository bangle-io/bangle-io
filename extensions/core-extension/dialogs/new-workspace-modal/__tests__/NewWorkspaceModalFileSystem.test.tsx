/**
 * @jest-environment jsdom
 */
import { act, fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';

import { useSerialOperationContext } from '@bangle.io/api';
import {
  BaseFileSystemError,
  NATIVE_BROWSER_PERMISSION_ERROR,
  pickADirectory,
} from '@bangle.io/baby-fs';
import { CORE_OPERATIONS_CREATE_NATIVE_FS_WORKSPACE } from '@bangle.io/constants';
import { useUIManagerContext } from '@bangle.io/slice-ui';
import { hasWorkspace } from '@bangle.io/slice-workspace';
import { OverlayProvider } from '@bangle.io/ui-components';

import { WORKSPACE_AUTH_REJECTED_ERROR } from '../common';
import { NewWorkspaceModal } from '../NewWorkspaceModal';

jest.mock('@react-aria/ssr/dist/main', () => {
  return {
    ...jest.requireActual('@react-aria/ssr/dist/main'),
    // react aria generates a bunch of random ids, this
    // makes the snapshot stable.
    useSSRSafeId: () => 'react-aria-test-id',
  };
});

jest.mock('@bangle.io/slice-ui', () => {
  const otherThings = jest.requireActual('@bangle.io/slice-ui');

  return {
    ...otherThings,
    useUIManagerContext: jest.fn(() => ({})),
  };
});

jest.mock('../common', () => {
  const otherThings = jest.requireActual('../common');

  return {
    ...otherThings,
    defaultStorageType: otherThings.FILE_SYSTEM,
  };
});

jest.mock('@bangle.io/baby-fs', () => {
  const otherThings = jest.requireActual('@bangle.io/baby-fs');

  return {
    supportsNativeBrowserFs: jest.fn(() => false),

    ...otherThings,
    pickADirectory: jest.fn().mockResolvedValue({}),
  };
});

jest.mock('react-dom', () => {
  const otherThings = jest.requireActual('react-dom');

  return {
    ...otherThings,
    createPortal: jest.fn((element, node) => {
      return element;
    }),
  };
});
jest.mock('@bangle.io/slice-workspace', () => {
  const workspaceThings = jest.requireActual('@bangle.io/slice-workspace');

  return {
    ...workspaceThings,
    hasWorkspace: jest.fn(() => () => {}),
  };
});

jest.mock('@bangle.io/api', () => {
  const otherThings = jest.requireActual('@bangle.io/api');

  return {
    ...otherThings,
    useSerialOperationContext: jest.fn(() => ({})),
  };
});

const hasWorkspaceMock = hasWorkspace as jest.MockedFunction<
  typeof hasWorkspace
>;

beforeEach(() => {
  let dispatchSerialOperation = jest.fn();

  hasWorkspaceMock.mockImplementation(() => async () => false);

  (useUIManagerContext as any).mockImplementation(() => {
    const dispatch = jest.fn();

    return {
      dispatch: dispatch,
    };
  });

  (pickADirectory as any).mockResolvedValue({});

  (useSerialOperationContext as any).mockImplementation(() => {
    return { dispatchSerialOperation };
  });
});

describe('NewWorkspaceModalFileSystem', () => {
  test('renders correctly', async () => {
    let result = render(
      <div>
        <OverlayProvider>
          <NewWorkspaceModal />
        </OverlayProvider>
      </div>,
    );
    expect(result.getByLabelText('Select storage type').innerHTML).toContain(
      'File System',
    );

    expect(result.container).toMatchSnapshot();
  });

  test('allows switching of storage type', async () => {
    let result = render(
      <div>
        <OverlayProvider>
          <NewWorkspaceModal />
        </OverlayProvider>
      </div>,
    );

    expect(result.getByLabelText('Select storage type').innerHTML).toContain(
      'File System',
    );

    act(() => {
      fireEvent.click(result.getByLabelText('Select storage type'));
    });

    await waitFor(() => {
      const dropdown = result.getByRole('listbox');
      expect(dropdown).toBeTruthy();
    });

    expect(result.getAllByRole('option').map((r) => r.getAttribute('data-key')))
      .toMatchInlineSnapshot(`
      Array [
        "file-system",
        "browser",
      ]
    `);

    const browserOpt = result
      .getAllByRole('option')
      .find((r) => r.getAttribute('data-key') === 'browser');

    fireEvent.click(browserOpt!);

    expect(result.getByLabelText('Select storage type').innerHTML).toContain(
      'Browser',
    );
  });

  test('is able create a workspace', async () => {
    let dispatchSerialOperation = jest.fn();
    (useSerialOperationContext as any).mockImplementation(() => {
      return { dispatchSerialOperation };
    });

    (pickADirectory as any).mockResolvedValue({ name: 'test-dir-name' });

    let result = render(
      <div>
        <OverlayProvider>
          <NewWorkspaceModal />
        </OverlayProvider>
      </div>,
    );

    expect(
      result.getByLabelText('Create workspace').hasAttribute('disabled'),
    ).toBe(true);

    act(() => {
      fireEvent.click(result.getByLabelText('pick directory'));
    });

    await waitFor(() => {
      const dropdown = result.getByLabelText('pick directory');

      expect(dropdown.innerHTML).toContain('test-dir-name');
    });

    expect(pickADirectory).toBeCalledTimes(1);

    expect(
      result.getByLabelText('Create workspace').hasAttribute('disabled'),
    ).toBe(false);

    act(() => {
      fireEvent.click(result.getByLabelText('Create workspace'));
    });

    expect(dispatchSerialOperation).toBeCalledTimes(1);
    expect(dispatchSerialOperation).nthCalledWith(1, {
      name: CORE_OPERATIONS_CREATE_NATIVE_FS_WORKSPACE,
      value: {
        rootDirHandle: {
          name: 'test-dir-name',
        },
      },
    });
  });

  test('handles when fs throws error', async () => {
    let dispatchSerialOperation = jest.fn();
    (useSerialOperationContext as any).mockImplementation(() => {
      return { dispatchSerialOperation };
    });

    let error = new BaseFileSystemError({
      message: `Permission rejected `,
      code: NATIVE_BROWSER_PERMISSION_ERROR,
    });

    (pickADirectory as any).mockRejectedValue(error);

    let result = render(
      <div>
        <OverlayProvider>
          <NewWorkspaceModal />
        </OverlayProvider>
      </div>,
    );

    act(() => {
      fireEvent.click(result.getByLabelText('pick directory'));
    });

    await waitFor(() => {
      return result.getByTestId(WORKSPACE_AUTH_REJECTED_ERROR);
    });

    expect(result.getByTestId(WORKSPACE_AUTH_REJECTED_ERROR))
      .toMatchInlineSnapshot(`
        <div
          class="w-full px-4 text-center rounded"
          data-testid="WORKSPACE_AUTH_REJECTED"
          style="color: white;"
        >
          <div
            class="font-semibold text-left"
          >
            Bangle.io was denied access to your notes.
          </div>
          <div
            class="text-left"
          >
            <div>
              Please try again and press allow Bangle.io access to your locally saved notes.
            </div>
          </div>
        </div>
      `);

    expect(
      result.getByLabelText('Create workspace').hasAttribute('disabled'),
    ).toBe(true);
  });

  test('is able to resume after fs error', async () => {
    let dispatchSerialOperation = jest.fn();
    (useSerialOperationContext as any).mockImplementation(() => {
      return { dispatchSerialOperation };
    });

    let error = new BaseFileSystemError({
      message: `Permission rejected `,
      code: NATIVE_BROWSER_PERMISSION_ERROR,
    });

    (pickADirectory as any)
      .mockRejectedValueOnce(error)
      .mockResolvedValueOnce({ name: 'test-dir-name' });

    let result = render(
      <div>
        <OverlayProvider>
          <NewWorkspaceModal />
        </OverlayProvider>
      </div>,
    );

    act(() => {
      fireEvent.click(result.getByLabelText('pick directory'));
    });

    await waitFor(() => {
      return result.getByTestId(WORKSPACE_AUTH_REJECTED_ERROR);
    });

    expect(pickADirectory).toBeCalledTimes(1);

    expect(
      result.getByLabelText('Create workspace').hasAttribute('disabled'),
    ).toBe(true);

    act(() => {
      fireEvent.click(result.getByLabelText('pick directory'));
    });

    await waitFor(() => {
      return result
        .getByLabelText('pick directory')
        .innerHTML.includes('test-dir-name');
    });

    expect(pickADirectory).toBeCalledTimes(2);

    expect(
      result.getByLabelText('Create workspace').hasAttribute('disabled'),
    ).toBe(false);
  });
});
