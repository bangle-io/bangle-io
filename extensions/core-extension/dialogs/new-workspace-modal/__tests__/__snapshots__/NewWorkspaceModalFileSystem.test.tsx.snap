// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewWorkspaceModalFileSystem renders correctly 1`] = `
<div>
  <div>
    <div
      aria-labelledby="modal-title"
      aria-modal="true"
      class="fixed z-50 inset-0 overflow-y-auto"
      role="dialog"
    >
      <div
        class="flex justify-center min-h-screen pb-20 text-center sm:block sm:p-0"
      >
        <div
          aria-hidden="true"
          class="fixed w-full h-full transition-opacity B-ui-components_modal-overlay"
        />
        <span
          aria-hidden="true"
          class="hidden sm:inline-block sm:align-middle sm:h-screen"
        >
          â€‹
        </span>
        <div
          class="B-ui-components_modal-container h-full B-ui-components_modal-fadeInScaleAnimation  inline-block align-bottom rounded-lg text-left overflow-hidden shadow-xl transform transition-all sm:my-8 sm:align-middle max-w-2xl "
          style="width: 30rem; max-width: 30rem;"
        >
          <div
            class="w-full px-6 my-2 select-none "
          >
            <div
              class="flex flex-row justify-between pt-2 pb-2 text-3xl border-b-2 B-ui-components_modal-header "
            >
              <span>
                New Workspace
              </span>
              <button
                class="focus:outline-none"
                data-bangle-editor-break="true"
                data-bangle-editor-pos="bottom"
                type="button"
              >
                <svg
                  class="fill-current  w-6 h-6 rounded-sm"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M6.2253 4.81108C5.83477 4.42056 5.20161 4.42056 4.81108 4.81108C4.42056 5.20161 4.42056 5.83477 4.81108 6.2253L10.5858 12L4.81114 17.7747C4.42062 18.1652 4.42062 18.7984 4.81114 19.1889C5.20167 19.5794 5.83483 19.5794 6.22535 19.1889L12 13.4142L17.7747 19.1889C18.1652 19.5794 18.7984 19.5794 19.1889 19.1889C19.5794 18.7984 19.5794 18.1652 19.1889 17.7747L13.4142 12L19.189 6.2253C19.5795 5.83477 19.5795 5.20161 19.189 4.81108C18.7985 4.42056 18.1653 4.42056 17.7748 4.81108L12 10.5858L6.2253 4.81108Z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class=""
          >
            <div
              class="px-6 py-4 select-none"
            >
              <div
                class="flex flex-col mb-5"
              >
                <div
                  class="mb-2"
                >
                  <h2
                    class="text-lg font-medium"
                  >
                    Storage type
                  </h2>
                  <span
                    class="text-sm"
                  >
                    Bangle.io strongly recommends you choose 
                    <b>
                      File system
                    </b>
                     
                    storage because it is safer and portable. Read more at
                     
                    <a
                      class="underline"
                      href="https://bangle.io/"
                    >
                      https://bangle.io
                    </a>
                    .
                  </span>
                </div>
                <div>
                  <div
                    class="inline-flex flex-col relative mt-4 w-64"
                  >
                    <div
                      class="B-ui-components_select-label text-sm font-medium cursor-default"
                    />
                    <div
                      aria-hidden="true"
                      style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
                    >
                      <input
                        style="font-size: 16px;"
                        tabindex="-1"
                        type="text"
                      />
                      <label>
                        <select
                          size="2"
                          tabindex="-1"
                        >
                          <option />
                          <option
                            value="file-system"
                          />
                          <option
                            value="browser"
                          >
                            Browser
                          </option>
                        </select>
                      </label>
                    </div>
                    <button
                      aria-describedby="react-aria-test-id react-aria-test-id"
                      aria-expanded="false"
                      aria-haspopup="listbox"
                      aria-label="Select storage type"
                      aria-labelledby="react-aria-test-id react-aria-test-id"
                      class="B-ui-components_select-button p-1 relative inline-flex flex-row items-center justify-between overflow-hidden cursor-default shadow-sm outline-none"
                      id="react-aria-test-id"
                      style="border-radius: var(--BV-ui-bangle-button-radius);"
                      type="button"
                    >
                      <span
                        class="text-md text-left px-1 w-64 truncate overflow-ellipsis "
                        id="react-aria-test-id"
                      >
                        File System
                         
                        (Only on Chromium browsers)
                      </span>
                      <svg
                        aria-hidden="true"
                        class="w-5 h-5"
                        fill="currentColor"
                        viewBox="0 0 20 20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          clip-rule="evenodd"
                          d="M10 3a1 1 0 01.707.293l3 3a1 1 0 01-1.414 1.414L10 5.414 7.707 7.707a1 1 0 01-1.414-1.414l3-3A1 1 0 0110 3zm-3.707 9.293a1 1 0 011.414 0L10 14.586l2.293-2.293a1 1 0 011.414 1.414l-3 3a1 1 0 01-1.414 0l-3-3a1 1 0 010-1.414z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="flex flex-col mb-5"
              >
                <div
                  class="mb-2"
                >
                  <h2
                    class="text-lg font-medium"
                  >
                    Location
                  </h2>
                  <span
                    class="text-sm"
                  >
                    Select a folder where Bangle.io will put all your notes. You can use an existing folder which has Markdown notes or create a new folder.
                  </span>
                </div>
                <div>
                  <button
                    aria-label="pick directory"
                    class="B-ui-bangle-button_button p-1  transition-all duration-100 B-ui-components_misc-button-ring"
                    id="new-workspace-modal_browse-button"
                    type="button"
                  >
                    <span
                      class=""
                    >
                      Browse
                    </span>
                  </button>
                </div>
              </div>
              <div
                class="flex flex-row justify-center"
              >
                <button
                  aria-label="create workspace"
                  class="px-4 B-ui-bangle-button_button p-1  transition-all duration-100 cursor-not-allowed"
                  disabled=""
                  id="new-workspace-modal_create-button"
                  type="button"
                >
                  <span
                    class=""
                  >
                    Create Workspace
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
